pygem/cffd.py

```diff
@@ -0,0 +1,144 @@
+
+
+class CFFD(FFD):
+    """
+    Class that handles the Constrained Free Form Deformation on the mesh points.
+
+    :param list n_control_points: number of control points in the x, y, and z
+        direction. Default is [2, 2, 2].
+        
+    :cvar numpy.ndarray box_length: dimension of the FFD bounding box, in the
+        x, y and z direction (local coordinate system).
+    :cvar numpy.ndarray box_origin: the x, y and z coordinates of the origin of
+        the FFD bounding box.
+    :cvar numpy.ndarray rot_angle: rotation angle around x, y and z axis of the
+        FFD bounding box.
+    :cvar numpy.ndarray n_control_points: the number of control points in the
+        x, y, and z direction.
+    :cvar numpy.ndarray array_mu_x: collects the displacements (weights) along
+        x, normalized with the box length x.
+    :cvar numpy.ndarray array_mu_y: collects the displacements (weights) along
+        y, normalized with the box length y.
+    :cvar numpy.ndarray array_mu_z: collects the displacements (weights) along
+        z, normalized with the box length z.
+    :cvar callable linconstraint: it defines the F of the constraint F(x)=c.
+    :cvar numpy.ndarray valconstraint: it defines the c of the constraint F(x)=c.
+    :cvar list indices: it defines the indices of the control points 
+        that are moved to enforce the constraint. The control index is obtained by doing:
+        all_indices=np.arange(n_x*n_y*n_z*3).reshape(n_x,n_y,n_z,3).tolist().
+    :cvar numpy.ndarray M: a SDP weigth matrix. It must be of size len(indices) x len(indices).
+
+    :Example:
+
+        >>> from pygem import CFFD
+        >>> import numpy as np
+        >>> cffd = CFFD()
+        >>> cffd.read_parameters('tests/test_datasets/parameters_test_ffd_sphere.prm')
+        >>> original_mesh_points = np.load('tests/test_datasets/meshpoints_sphere_orig.npy')
+        >>> A=np.random.rand(3,original_mesh_points.reshape(-1).shape[0])
+        >>> def fun(x):
+        >>>     x=x.reshape(-1)
+        >>>     return A@x
+        >>> b=fun(original_mesh_points)
+        >>> cffd.linconstraint=fun
+        >>> cffd.valconstraint=b
+        >>> cffd.indices=np.arange(np.prod(cffd.n_control_points)*3).tolist()
+        >>> cffd.M=np.eye(len(cffd.indices))
+        >>> new_mesh_points = cffd(original_mesh_points)
+        >>> assert np.isclose(np.linalg.norm(fun(new_mesh_points)-b),np.array([0.]))
+    """
+
+    def __init__(self, n_control_points=None):
+        super().__init__(n_control_points)
+        self.linconstraint = None 
+        self.valconstraint = None
+        self.indices = None
+        self.M = None
+
+    def __call__(self, src_pts):
+        saved_parameters=self._save_parameters()
+        A, b = self._compute_linear_map(src_pts, saved_parameters.copy())
+        d = A @ saved_parameters[self.indices] + b
+        deltax = np.linalg.inv(self.M) @ A.T @ np.linalg.inv((A @ np.linalg.inv(self.M)@ A.T)) @ (self.valconstraint - d)
```

Comment:
You're computing the inverse of M twice in same line. Please, save intermediate result in a variable